csharp/src/Google.Protobuf/UnknownFieldSet.cs

```diff
@@ -0,0 +1,364 @@
+            {
+                return true;
+            }
+            UnknownFieldSet otherSet = other as UnknownFieldSet;
+            IDictionary<int, UnknownField> otherFields = otherSet.fields;
+            if (fields.Count  != otherFields.Count)
+            {
+                return false;
+            }
+            foreach (KeyValuePair<int, UnknownField> leftEntry in fields)
+            {
+                UnknownField rightValue;
+                if (!otherFields.TryGetValue(leftEntry.Key, out rightValue))
+                {
+                    return false;
+                }
+                if (!leftEntry.Value.Equals(rightValue))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Gets the unknown field set's hash code.
+        /// </summary>
+        public override int GetHashCode()
+        {
+            int ret = 1;
+            foreach (KeyValuePair<int, UnknownField> field in fields)
+            {
+                int hash = field.Key.GetHashCode() ^ field.Value.GetHashCode();
+                ret ^= hash;
+            }
+            return ret;
+        }
+
+        #region IMessage Members
+        /// <summary>
+        /// Gets the descriptor. It is never used. Returns null to implement
+        /// the interface member 'IMessage.Descriptor'.
+        /// </summary>
+        public MessageDescriptor Descriptor
+        {
+            get { return null; }
+        }
+
+        /// <summary>
+        /// Checks if the unknown field set is initialized. It is never used.
+        /// Returns true to implement the interface.
+        /// </summary>
+        public bool IsInitialized
+        {
+            get { return true; }
+        }
+        #endregion
+
+        // Optimization:  We keep around the last field that was
+        // modified so that we can efficiently add to it multiple times in a
+        // row (important when parsing an unknown repeated field).
+        private int lastFieldNumber;
+        private UnknownField lastField;
+
+        private UnknownField GetOrAddField(int number)
+        {
+            if (lastField != null && number == lastFieldNumber)
+            {
+                return lastField;
+            }
+            if (number == 0)
+            {
+                return null;
+            }
+
+            UnknownField existing;
+            if (fields.TryGetValue(number, out existing))
+            {
+                return existing;
+            }
+            lastField = new UnknownField();
+            AddField(number, lastField);
+            lastFieldNumber = number;
+            return lastField;
+        }
+
+        /// <summary>
+        /// Adds a field to the set. If a field with the same number already exists, it
+        /// is replaced.
+        /// </summary>
+        public UnknownFieldSet AddField(int number, UnknownField field)
```

Comment:
Calling this AddField is confusing when .NET dictionaries tend to throw if you try to add an existing key.

AddOrReplaceField perhaps?

Additionally, it looks like this could be internal (even then, only visible for testing.)